### PR TITLE
Implement kubeadm reset

### PR DESCRIPTION
**Pull request by luxas**
@kubernetes/sig-cluster-lifecycle

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34807)

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/init.go]: Intentional?

File: cmd/kubeadm/app/preflight/checks.go
Comment (dgoodwin) [inline cmd/kubeadm/app/preflight/checks.go]: Did these turn out to not be necessary?

File: cmd/kubeadm/app/cmd/reset.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/reset.go]: This shows up as: Deleting the stateful directories: [[/var/lib/kubelet /var/lib/etcd /etc/kubernetes]]

Can probably drop the [] in code.

File: cmd/kubeadm/app/cmd/reset.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/reset.go]: Should we actually clean these up or just kill? My reset script I'd been clearing them out entirely.

Also should we only hit containers with a k8s prefix?

**Review by dgoodwin — COMMENTED** [review]
This worked great on CentOS, just a few questions and one very minor request for changes.

File: cmd/kubeadm/app/cmd/reset.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/reset.go]: Thinking we may want to filter to k8s containers only, I could envision a developer deciding to init master on their workstation.

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Yep, I think we decided to drop it
At least it was dropped in join

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: I thought about that, leads to more work, but is probably worth it

Not sure if we should remove them though... the user might want to read logs

File: cmd/kubeadm/app/cmd/init.go
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: Drop it.

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: This is for the node, 10251 and 10252 for cm and scheduler and 8080 for unsecured apiserver

File: cmd/kubeadm/app/cmd/reset.go
Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/reset.go]: Why doesn't `initsystem.GetInitSystem()` return that string as an error?

File: cmd/kubeadm/app/cmd/reset.go
Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/reset.go]: I think `find /var/lib/kubelet | xargs -n 1 findmnt -n -t tmpfs -o TARGET -T | uniq | xargs -r umount ` is cleaner.

Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/reset.go]: If you depend on `awk`, you don't have to use `grep`, i.e. `awk '$2 ~ /\/var\/lib\/kubelet { print $2 }' | xargs umount`, or even `awk '$2 ~ /\/var\/lib\/kubelet { system("umount "$2) }`, on the other hand you don't even have to shell out...

File: cmd/kubeadm/app/cmd/reset.go
Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/reset.go]: I think we should use `docker rm --force --volumes` here.

> the user might want to read logs

Maybe, but more user will complain about disk space used up by those containers.

**Review by errordeveloper — CHANGES_REQUESTED** [review]

**Comment by luxas** [comment]
PTAL @pires @mikedanese @dgoodwin @errordeveloper

**Comment by luxas** [comment]
I'd like to get this lgtm'd ASAP, if @mikedanese could do it, it would be awesome! (Ilya is offline now)

**Review by dgoodwin — APPROVED** [review]
LGTM

**Comment by luxas** [comment]
Applying @dgoodwin's LGTM

File: cmd/kubeadm/app/cmd/reset.go
Comment (pires) [inline cmd/kubeadm/app/cmd/reset.go]: Why the empty line?

File: cmd/kubeadm/app/cmd/reset.go
Comment (pires) [inline cmd/kubeadm/app/cmd/reset.go]: This will break pre-flight checks. It should be `return &preflight.PreFlightError{Msg: err.Error()}}` instead.

**Review by pires — CHANGES_REQUESTED** [review]

**Comment by luxas** [comment]
Fixed @pires' comments, waiting for the build to be green before applying lgtm again

**Review by pires — APPROVED** [review]
LGTM

**Pull request merged**
